cpp/daal/src/algorithms/dtrees/gbt/treeshap.h

```diff
@@ -0,0 +1,476 @@
+float unwoundPathSum(const PathElement * uniquePath, const float * pWeights, unsigned uniqueDepth, unsigned uniqueDepthPWeights, unsigned pathIndex);
+float unwoundPathSumZero(const float * pWeights, unsigned uniqueDepth, unsigned uniqueDepthPWeights);
+
+/**
+ * Recursive Fast TreeSHAP version 1
+ * Important: nodeIndex is counted from 0 here!
+*/
+template <typename algorithmFPType, bool hasUnorderedFeatures, bool hasAnyMissing>
+inline void treeShap(const gbt::internal::GbtDecisionTree * tree, const algorithmFPType * x, algorithmFPType * phi,
+                     const FeatureTypes * featureHelper, size_t nodeIndex, size_t depth, size_t uniqueDepth, size_t uniqueDepthPWeights,
+                     PathElement * parentUniquePath, float * parentPWeights, algorithmFPType pWeightsResidual, float parentZeroFraction,
+                     float parentOneFraction, int parentFeatureIndex, int condition, FeatureIndexType conditionFeature, float conditionFraction)
+{
+    // stop if we have no weight coming down to us
+    if (conditionFraction < FLT_EPSILON) return;
+
+    const size_t numOutputs                   = 1; // currently only support single-output models
+    const ModelFPType * const splitValues     = tree->getSplitPoints() - 1;
+    const int * const defaultLeft             = tree->getDefaultLeftForSplit() - 1;
+    const FeatureIndexType * const fIndexes   = tree->getFeatureIndexesForSplit() - 1;
+    const ModelFPType * const nodeCoverValues = tree->getNodeCoverValues() - 1;
+
+    // extend the unique path
+    PathElement * uniquePath = parentUniquePath + uniqueDepth + 1;
+    size_t nBytes            = (uniqueDepth + 1) * sizeof(PathElement);
+    int copyStatus           = daal::services::internal::daal_memcpy_s(uniquePath, nBytes, parentUniquePath, nBytes);
+    DAAL_ASSERT(copyStatus == 0);
+    // extend pWeights
+    float * pWeights = parentPWeights + uniqueDepthPWeights + 1;
+    nBytes           = (uniqueDepthPWeights + 1) * sizeof(float);
+    copyStatus       = daal::services::internal::daal_memcpy_s(pWeights, nBytes, parentPWeights, nBytes);
+    DAAL_ASSERT(copyStatus == 0);
+
+    if (condition == 0 || conditionFeature != static_cast<unsigned>(parentFeatureIndex))
+    {
+        extendPath(uniquePath, pWeights, uniqueDepth, uniqueDepthPWeights, parentZeroFraction, parentOneFraction, parentFeatureIndex);
+        // update pWeightsResidual if the feature of the last split does not satisfy the threshold
+        if (parentOneFraction != 1)
+        {
+            pWeightsResidual *= parentZeroFraction;
+            uniqueDepthPWeights -= 1;
+        }
+    }
+
+    const bool isLeaf = gbt::internal::ModelImpl::nodeIsLeaf(nodeIndex, *tree, depth);
+
+    if (isLeaf)
+    {
+        // +1 to account for -1 in splitValues array
+        const unsigned valuesOffset = nodeIndex * numOutputs;
+        unsigned valuesNonZeroInd   = 0;
+        unsigned valuesNonZeroCount = 0;
```

Comment:
Why don't you use `std::uint32_t`?

Reply:
Agree here with Nikita.
Even if the code is copied from some third party source, we need to align it with our [coding guidelines](https://oneapi-src.github.io/oneDAL/contribution/coding_guide.html).
Otherwise our codebase will evolve into a patchwork someday.